client/network/sync/src/engine.rs

```diff
@@ -590,6 +608,33 @@ where
 			self.tick_timeout.reset(TICK_TIMEOUT);
 		}
 
+		// go over all connected peers and check if any of them have been idle for a while. Idle in
+		// this case means that we haven't sent or received block announcements to/from this peer.
+		// If that is the case, because of #5685, it could be that the block announces substream is
+		// not actually open and and this peer is just wasting a slot and is should be replaced with
+		// some other node that is willing to send us block announcements.
+		for (id, peer) in self.peers.iter() {
+			// because of a delay between disconnecting a peer in `SyncingEngine` and getting the
+			// response back from `Protocol`, a peer might be reported and disconnect multiple
+			// times. To prevent this from happening (until the underlying issue is fixed), keep
+			// track of evicted peers and report and disconnect them only once.
+			if self.evicted.contains(id) {
```

Comment:
Does this mean that if the peer reconnects to us and don't send any new notifications, it won't be evicted again?

Reply:
We remove from evicted when the `NotificationStream` is closed below.